### PR TITLE
SUBMARINE-1023. Submarine SDK sqlalchemy store (model registry)

### PR DESCRIPTION
**Pull request by KUAN-HSUN-LI**
### What is this PR for?
* Implement the model registry SQL method in Python SDK
* Apply sqlalchemy mypy checks
* Replace submarine tracking_uri with db_uri

### What type of PR is it?
[Feature]

### What is the Jira issue?
https://issues.apache.org/jira/browse/SUBMARINE-1023

### How should this be tested?
All of the tests are provided in `submarine-sdk/pysubmarine/tests/store/model_registry/test_sqlalchemy_store.py`

### Questions:
* Do the license files need updating? No
* Are there breaking changes for older versions? No
* Does this need new documentation? No

### CODE REVIEW

**Comment by KUAN-HSUN-LI** [comment]
@pingsutw @jeff-901  Can you help me review the code? Thanks

File: submarine-sdk/pysubmarine/submarine/store/model_registry/sqlalchemy_store.py [hunk 1]
```diff
@@ -0,0 +1,570 @@
...
+        :param version: Registered model version.
+        :param description: New model description.
+        :return: A single :py:class:`submarine.entities.model_registry.ModelVersion` object.
+        """
+        with self.ManagedSessionMaker() as session:
+            update_time = datetime.now()
+            sql_model_version = self._get_model_version(session, name, version)
+            sql_model_version.description = description
+            sql_model_version.last_updated_time = update_time
+            self._save_to_db(session, sql_model_version)
```
Comment (jeff-901) [inline submarine-sdk/pysubmarine/submarine/store/model_registry/sqlalchemy_store.py:441]: Is it necessary to check the length of description?

File: submarine-sdk/pysubmarine/submarine/store/model_registry/sqlalchemy_store.py [hunk 2]
```diff
@@ -0,0 +1,570 @@
...
+        List of all model versions that satisfy the filter criteria.
+        :param name: Registered model name.
+        :param filter_tags: Filter tags, defaults not to filter any tags.
+        :return: A List of :py:class:`submarine.entities.model_registry.ModelVersion` objects
+                that satisfy the search expressions.
+        """
+        conditions = [SqlModelVersion.name == name]
+        if filter_tags is not None:
+            conditions = [
+                SqlModelVersion.model_tags.any(SqlModelTag.tag.contains(tag)) for tag in filter_tags
```
Comment (jeff-901) [inline submarine-sdk/pysubmarine/submarine/store/model_registry/sqlalchemy_store.py:501]: Is it wrong to reassign "conditions" when filter_tags is not None?

File: submarine-sdk/pysubmarine/tests/store/model_registry/test_sqlalchemy_store.py [hunk 3]
```diff
@@ -0,0 +1,731 @@
...
+        # empty result
+        other_tag = ["tag4"]
+        results = self.store.list_model_version(name, filter_tags=other_tag)
+        self.assertEqual(len(results), 0)
+        self.assertEqual(results, [])
+
+        # empty result
+        results = self.store.list_registered_model(filter_tags=tags + other_tag)
+        self.assertEqual(len(results), 0)
+        self.assertEqual(results, [])
```
Comment (jeff-901) [inline submarine-sdk/pysubmarine/tests/store/model_registry/test_sqlalchemy_store.py:613]: This test only contains one registered model name. Adding other registered model name helps to test list_model_version thoroughly when both name and tags are provided.

File: submarine-sdk/pysubmarine/submarine/store/model_registry/sqlalchemy_store.py (hunk 1 above)
Comment (KUAN-HSUN-LI) [reply submarine-sdk/pysubmarine/submarine/store/model_registry/sqlalchemy_store.py:441]: I will add the check

File: submarine-sdk/pysubmarine/submarine/store/model_registry/sqlalchemy_store.py (hunk 2 above)
Comment (KUAN-HSUN-LI) [reply submarine-sdk/pysubmarine/submarine/store/model_registry/sqlalchemy_store.py:501]: Thanks for point out this problem

File: submarine-sdk/pysubmarine/submarine/utils/validation.py [hunk 4]
```diff
@@ -116,6 +117,45 @@ def validate_param(key, value):
...
+        validate_tag(tag)
+
+
+def validate_tag(tag: str) -> None:
+    """Check that `tag` is a valid tag value and raise an exception if it isn't."""
+    # Reuse param & metric check.
+    if tag is None or tag == "":
+        raise SubmarineException("Tag cannot be empty.")
+    if not _VALID_PARAM_AND_METRIC_NAMES.match(tag):
+        raise SubmarineException("Invalid tag name: '%s'. %s" % (tag, _BAD_CHARACTERS_MESSAGE))
```
Comment (pingsutw) [inline submarine-sdk/pysubmarine/submarine/utils/validation.py:133]:
```suggestion
        raise SubmarineException(f"Invalid tag name: {tag}. {_BAD_CHARACTERS_MESSAGE}")
```

File: submarine-sdk/pysubmarine/github-actions/test-requirements.txt [hunk 5]
```diff
@@ -23,11 +23,12 @@ pytest==3.2.1
 pytest-cov==2.6.0
 pytest-localserver==0.5.0
 pylint==2.5.2
-sqlalchemy==1.3.0
+sqlalchemy >= 1.4.0
```
[the remark this answers is not in the document]
Comment (KUAN-HSUN-LI) [reply submarine-sdk/pysubmarine/github-actions/test-requirements.txt:26]: Sure, I will remove it.

File: submarine-sdk/pysubmarine/submarine/store/model_registry/abstract_store.py [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+
+    def __init__(self) -> None:
+        """
+        Empty constructor for now. This is deliberately not marked as abstract, else every
+        derived class would be forced to create one.
+        """
+        pass
+
+    @abstractmethod
+    def create_registered_model(
```
[the remark this answers is not in the document]
Comment (KUAN-HSUN-LI) [reply submarine-sdk/pysubmarine/submarine/store/model_registry/abstract_store.py:38]: model registry including `registered_model`, `registered_model_tag`, `model_version`, `model_version_tag` four tables.
tracking including `param`, `metric`, `experiment` three tables.
In my opinion, it is more clear to put these functions in different directories. In this PR, I only worked on the model registry directory maybe I should also refactor the tracking directory and implement the `experiment` table.

File: submarine-sdk/pysubmarine/submarine/store/model_registry/sqlalchemy_store.py [hunk 7]
```diff
@@ -0,0 +1,572 @@
...
+
+    def create_model_version(
+        self,
+        name: str,
+        source: str,
+        user_id: str,
+        experiment_id: str,
+        dataset: str = None,
+        description: str = None,
+        tags: List[str] = None,
```
[the remark this answers is not in the document]
Comment (KUAN-HSUN-LI) [reply submarine-sdk/pysubmarine/submarine/store/model_registry/sqlalchemy_store.py:348]: `registered model tag` focuses on what is the goal of this model or the category of this model. `model version tag` focuses on what this model perform or feature of this model. I think it is a little confused in these two parts. Hopefully, this explanation can help.

**Comment by KUAN-HSUN-LI** [comment]
@pingsutw @jeff-901 Thanks for your review. I have replaced `registered model` with `model container` and replaced `model version` with `model metadata`. Additionally, I have fixed serval bugs.

File: submarine-sdk/pysubmarine/submarine/utils/validation.py [hunk 8]
```diff
@@ -116,8 +117,48 @@ def validate_param(key, value):
...
+
+
+def validate_model_name(model_name: str) -> None:
+    if model_name is None or model_name == "":
+        raise SubmarineException("Registered model name cannot be empty.")
+
+
+def validate_model_version(model_version: int) -> None:
```
Comment (jeff-901) [inline submarine-sdk/pysubmarine/submarine/utils/validation.py:141]: model_version forget to change to model metadata

File: submarine-sdk/pysubmarine/tests/entities/model_registry/test_model_version.py [hunk 9]
```diff
@@ -15,12 +15,11 @@
 
 from datetime import datetime
```
Comment (jeff-901) [inline submarine-sdk/pysubmarine/tests/entities/model_registry/test_model_version.py:16]: This file name should be changed to test_model_metadata.py

File: submarine-sdk/pysubmarine/submarine/utils/validation.py (hunk 8 above)
Comment (KUAN-HSUN-LI) [reply submarine-sdk/pysubmarine/submarine/utils/validation.py:141]: I think this function name is fine.

Comment (jeff-901) [reply submarine-sdk/pysubmarine/submarine/utils/validation.py]: ok

**Review by jeff-901 — APPROVED** [review]
LGTM

**Comment by KUAN-HSUN-LI** [comment]
@pingsutw @jeff-901 I am so sorry for changing the table name in the database again. I think the previous names are the best. I will explain it as follow:
workflow of model registry:
1. save models in s3 bucket (unregistered model)
2. choose the best model and registry it (registered model, version 1)
3. train the model with different parameters or datasets and save models in s3 bucket (unregistered model)
4. Again, choose the best model and registry it (registered model, version 2)

Summary:
* The table is storing registered models so the name `registered_model` is more proper.
* The version of the registered model with the same name will define a different version which will be stored in another table. It is also proper to name the table `model_version`

We also need two tag tables.
example:
image

**Review by pingsutw — APPROVED** [review]

**Pull request closed without merging**